### ISSUE TITLE
Issues with type I error for CAR structures / possibility to rotate residuals?

### ISSUE THREAD

**Issue by florianhartig** [florianhartig/DHARMa#301]
Andrea Havron and Cole Monnahan alerted me to the fact that residual tests can have inflated Type I error when there is a residual correlation structure, and asked about the possibility of DHARMa residuals can be rotated, as currently implemented in TMB. 

First of all, here is an example, modified from an example sent by Andrea:

```
library(mvtnorm)
set.seed(1)
C <- exp(-as.matrix(dist(seq(0,50,by=.1))))
obs <- as.numeric(rmvnorm(1,sigma=C))
par(mfrow = c(1,2))
image(C, main = "Residual covariance")
plot(obs, type = "l")
```

<img width="842" alt="image" src="https://user-images.githubusercontent.com/5457753/128472326-209d26d6-1543-4231-bd37-5e89ed345fff.png">

Rotated analytical residuals and normal and rotated DHARMa Residuals

```
# TMB's 'one-step-ahead' method (See Thygesen et al. 2017 and TMB::oneStepPredict).
## L * L^T = C
L <- t(chol(C))
r1 <- solve(L, obs)
qqnorm(r1, main="OSA")
abline(0,1)
ks.test(r1, pnorm)

## DHARMA residuals:
x = replicate(100, as.numeric(rmvnorm(1,sigma=C)))

# calculate DHARMa residuals
library(DHARMa)
res <- createDHARMa(x, obs, integerResponse = F)
testUniformity(res)

# calculated rotated DHARMa residuals
x1 = apply(x, 2, function(a) solve(L, a))
res <- createDHARMa(x1, r1, integerResponse = F)
testUniformity(res)
```

Response from me: First of all, yes, CAR type residual correlations will prevail in DHARMa residuals IF residual simulations are performed unconditional, and these residual patterns can then incorrectly be flagged as problems. That correlations were not removed was already noted in the help of testSpatialAutocorrelation and testTemporalAutocorrelation, but the consequences were not explained clearly enough. This should be changed. 

Afaiks, there are 2 remedies for the problem: 

1. Given that the issue arises through unconditional simulations, one fix would be to perform simulations conditional on the fitted REs, if that is possible (it depends on the package, in lme4 it would be possible, but lme4 doesn't have CAR structures; glmmTMB has CAR structures but conditional simulations are (I think) not implemented as of today), but check https://rdrr.io/cran/glmmTMB/man/simulate.glmmTMB.html in the future. It should be possible in all Bayesian packages or TMB.

2. Alternatively, the proposed rotation could be implemented in DHARMa residuals. The problem, however, is that this will likely only work perfectly for symmetric linear homogenous cdfc, such as the multivariate normal. I can't see how we could find a unique rotation that would apply across all data in, e.g., s Poisson GLMM with a CAR structure. Nevertheless, an imperfect rotation may still be preferable over none at all. 

Proposed remedies: 

a) improve help, and alert users to the fact that they may want to consider conditional simulations for strong CAR structures

b) implement the option to supply a user-defined rotation in simulateResiduals. Additionally, one could also use the covariance of the simulated residuals to generate an approximate rotation, which could also work for GLMMs etc.

**Comment by florianhartig** [comment]
There is an experimental fix for this in branch https://github.com/florianhartig/DHARMa/tree/issue301-rotation. To install this, follow instructions on the https://github.com/florianhartig/DHARMa, how to install a dev branch

Changes in this branch:

- Added rotation option to all functions that create residuals (simulateResiduals, recalculateResiduals, createDHARMa)
- Added help comments about autocorrelation structures, in particular in simulateResiduals, testSpatialAutocorrelation, testTemporalAutocorrelation
- Added example about the use of rotation in testTemporalAutocorrelation

**Comment by Cole-Monnahan-NOAA** [comment]
CC @Andrea-Havron-NOAA

We will test this approach and report back. 

Your point about the imperfect rotation is a good one. That is the motivation behind the OSA methods like "CDF" but it is not really clear how imperfect it needs to be to matter. I agree that an imperfect rotation is probably better. But we should explore this.

**Comment by Cole-Monnahan-NOAA** [comment]
Just one question. Looking at this example

https://github.com/florianhartig/DHARMa/blob/issue301-rotation/DHARMa/inst/examples/testTemporalAutocorrelationHelp.R#L56

and the implementation at 

https://github.com/florianhartig/DHARMa/blob/c4daac80f46b79efb4b414b0bcef1a5ce27cf8d7/DHARMa/R/helper.R#L90

I thought you were going to implement an option like `rotation=TRUE` which then *internally* calculates the empirical covariance from the simulated observations like

````
C <- cov(simulations)
L <- t(chol(C))
observed <- solve(L, observed)
simulations = apply(simulations, 2, function(a) solve(L, a))
````

**Comment by florianhartig** [comment]
Hi Cole,

well, I thought I would start with your current option, but I have added an empirical estimate of the covariance now, the option is rotation = "estimated". My first tests with your example suggest that this works in principle, but 

b) I think it requires quite a few simulations to get sufficient accuracy (100 in your example didn't suffice, with 1000 it at least removed the autocorrelation, but there was a remaining residual pattern)

a) this is all getting super slow for high-dimensional data

Maybe things could be sped up by implementing a sparse version of this code. In the end, the question seems to be how can we get a fast estimate of the expected residual covariance of a fitted model. Possibly, it would be more efficient to let the user implement this estimate than forcing this on the DHARMa side.

**Comment by florianhartig** [comment]
OK, I have tested this again, seems to work perfectly within the discussed limitations - I will merge this into the master and include it in the next DHARMa update.

Feel free to add further comments / suggestions to this thread though!

**Comment by Cole-Monnahan-NOAA** [comment]
Hi @florianhartig apologies for the long delay. I did have a chance to test it out. I ran it through some simulated GLMMs with and without the rotation. I did a normality test on the residuals and found some differences in Type I errors in the distribution of p-values when the correctly-specified model was used. I would think that in the case where no rotation was needed it wouldn't hurt? I haven't had time to expand this investigation and can't say I fully understand the behavior. 

But I do think that it's programmed correctly and should be available to users as an (experimental) option. @Andrea-Havron-NOAA and I will try to explore this a bit more and let you know if we have any further feedback.

Thanks for the quick response and implementation. I can see this being a really useful feature in a subset of users cases.
